### ISSUE TITLE
Learned Policy Stage 1 — PolicyInput-only feature / tensor schemaをexperiment-localに確立する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-arena#129]
Predecessor: [`lisbun/lisjong #149`](https://github.com/lisbun/lisjong/issues/149)

Related planning:

- [`lisbun/lisjong-project #36`](https://github.com/lisbun/lisjong-project/issues/36)
- [`lisbun/lisjong-project #37`](https://github.com/lisbun/lisjong-project/issues/37)

# Summary

`lisjong #149`で、Learned Policy向けのversioned action vocabulary、`InternalAction` codec、legal maskが確立された。

次段階として、current `PolicyInput`だけから、最初のfeed-forward Learned Policy実験が利用できるfixed-size feature / tensor representationを構築する。

本Issueでは、representation choiceを十分な実測前に`lisjong`の恒久的なproduction contractへ昇格させない。`lisjong-arena`所有の **experiment-local, versioned, PolicyInput-only schema** として実装し、後続の最小behavior-cloning vertical sliceが再現可能に利用できる状態までをscopeとする。

```text
PolicyInput
    ↓
experiment-local semantic feature
    ↓
fixed-size float tuple / tensor adapter

DecisionContext.legal_actions
    ↓
lisjong action vocabulary legal mask   # separate path
```

model、training、teacher data生成、action selectionは本Issueでは実装しない。

---

# Why Arena-owned / experiment-local

`PolicyInput`は安定したseat-safe domain contractだが、そのどの情報をどのshape・normalizationでmodelへ渡すかは実験上の選択である。

最初の実測前にfeature layoutを`lisjong`の汎用契約へ固定すると、後続model familyや学習結果に応じた変更を不必要にbreaking changeへしてしまう。

Arenaには既にHandBelief研究用の `phase6-history-snapshot-v1` / 919次元schemaが存在するが、これは、

```text
FrozenPlayerSafeAnchor
+ ordered RoundEvidence prefix
```

を入力とするHandBelief estimator専用表現である。

本Issueではこれを、

- PolicyInput encoderのbase class
- canonical Learned Policy input
- compatibility contract
- 流用すべきtensor layout

として扱わない。

必要な実装patternは参照してよいが、feature semanticsとversion identityは独立させる。

---

# Goal

current `PolicyInput`を唯一のonline-side inputとして、以下を確立する。

1. fixed-sizeかつdeterministicなsemantic feature
2. versioned tensor layout
3. pure-Pythonのflat numeric representation
4. optional / lazyなtorch tensor adapter
5. feature groupごとのoffset、length、shape、値域、normalization
6. same-version内のnumeric assignmentを固定するcompatibility test
7. seat rotation、tile ordering、sequence、paddingの明示的なsemantics
8. malformed / out-of-scope inputのfail-closed behavior

後続Issueが、

```text
PolicyInput
    -> features
    -> model logits
    -> legal mask
    -> selected vocabulary index
    -> canonical InternalAction
```

を組み立てられることを目的とするが、そのend-to-end model path自体は本Issueのscope外とする。

---

# Input boundary

encoderが受け取れるonline-side inputは、exact typeのcurrent `PolicyInput`だけとする。

最低限、current contractの以下を棚卸しする。

```text
PolicyInput
├── self_seat
├── round: RoundState
│   ├── round_wind
│   ├── hand_number
│   ├── dealer_seat
│   ├── honba
│   ├── riichi_sticks
│   ├── dora_indicators
│   └── live_wall_tiles_remaining
├── players[4]: PlayerPublicState
│   ├── score
│   ├── discards
│   │   ├── tile
│   │   ├── tsumogiri
│   │   ├── order
│   │   └── called_by
│   ├── melds
│   │   ├── kind
│   │   ├── tiles
│   │   ├── from_seat
│   │   └── called_tile
│   └── riichi
└── own_hand: OwnHandState
    ├── concealed_tiles
    └── drawn_tile
```

次をencoderへ渡さない。

- `DecisionContext.legal_actions`
- legal mask
- selected Action
- MatchState
- true opponent concealed hands
- live / dead wall truth
- future state / outcome
- raw RiichiEnv Observation
- raw mjai event / full log
- Arena match runner
- HandBelief prediction
- policy-specific heuristic result

legal maskは`lisjong #149`の既存action-vocabulary APIから別経路で生成する。legal action compositionをfeatureへ重複格納しない。

---

# Preflight design gate

実装前にcurrent mainと適用instructionsを再取得する。

最低限確認するもの:

- this Issue本文と最新comments
- `lisbun/lisjong #149` final contract / merged implementation
- `lisjong-arena` root/path `AGENTS.md`
- `lisjong-arena` current `pyproject.toml` / CI
- current pinned `lisjong` revision
- `docs/architecture.md`
- current `phase05_belief_slice` feature implementation
- current `phase6_snapshot/feature.py`
- current `phase6_snapshot/tensor.py`
- `lisjong/docs/policy-input-schema.md`
- all `PolicyInput` component types and validation
- existing test fixture/builders for `PolicyInput`

実装前に、全feature groupについて次を表でlockする。

| Field | Required definition |
|---|---|
| source | 元となるexact `PolicyInput` field |
| visibility | player-safeである根拠 |
| representation | one-hot / count / scalar / ordered slots等 |
| logical shape | flatten前のshape |
| flat range | half-open index range |
| value domain | encoding前後の値域 |
| normalization | scaleと根拠 |
| seat semantics | absolute / self-relative |
| tile semantics | 34-type / 37-tile、赤牌の扱い |
| order semantics | preserved / canonicalized |
| missing semantics | presence bit / padding |
| maximum length | 固定上限とその根拠 |
| failure | 表現不能時の例外 |

併せて以下を確定する。

- feature semantics ID
- tensor schema version
- total dimension
- feature group order
- flatten order
- dtype
- exact tile axis
- exact seat-relative axis
- padding representation
- presence / padding mask
- normalization constants
- compatibility fingerprint input
- unsupported version / malformed inputの例外階層

Issue作成時点でdimensionを推測して固定しない。current codeを正本としてpreflightで確定し、implementation・tests・documentationに同じliteralを反映する。

---

# Design constraints

## 1. PolicyInput-only

同じ`PolicyInput` valueからは、execution environmentやlegal action列挙順に関係なく同じfeatureを生成する。

encoder内部からexternal stateを検索・参照しない。

## 2. Seat-relative representation

playerごとの情報は原則としてself-relative順に配置する。

```text
relative seat = (absolute seat - self_seat) mod 4
```

少なくとも、

- self
- shimocha
- toimen
- kamicha

に対応するaxisの意味を明記する。

dealer、`Discard.called_by`、`PublicMeld.from_seat`も同じmappingを使用する。

同じsemantic positionをabsolute seatだけ回転させたfixtureで、relative featureが一致することをtestする。

場風・自風などseat rotation後も意味が変わる情報は、意味を失わない独立featureとして表現する。

## 3. Tile semantics

tile axisとred-five handlingを明示し、same version内で固定する。

- tile type countだけでは赤5情報を黙って失わない
- red distinctionを独立channelにするか37-tile axisにするかをpreflightで決定する
- concealed tiles / meld tiles等のunordered multisetをinput tuple orderへ依存させない
- dora indicators、discards等の意味あるsequenceを一律sortしない

## 4. Ordered public history

`Discard.order`が表す全seat共通の時間順序を失う場合、そのlossを意図的なv1 design decisionとして明記し、後続BCへの影響を説明する。

sequenceを保持する場合は、

- global ordering
- per-seat slot placement
- maximum length
- padding
- presence
- `called_by`
- `tsumogiri`

を一意に定義する。

0値とpadding / not-applicableを同じ意味へ潰さない。

## 5. Variable-length fields

以下を含むvariable-length fieldをsilent truncateしない。

- dora indicators
- discards
- melds
- concealed tiles

fixed upper boundを採る場合は、current supported 4-player rulesまたは物理制約から根拠を示し、boundary testを追加する。

contract上validだが実験schemaの上限を超える`PolicyInput`は、末尾を捨てたりclipしたりせずfail closedとする。

## 6. Derived features

v1は原則として、current `PolicyInput`の基礎観測を構造的にencodeする。

以下を追加しない。

- shanten
- ukeire
- hand value
- expected value
- wait prediction
- danger / deal-in score
- HandBelief
- policy-specific ranking
- teacher Action

tile counts、remaining public inventory、seat rotation等を追加する場合は、「単なる構造変換」か「derived feature」かを明記する。既存値から一意に導けないfeatureは含めない。

## 7. Model independence

最初のconsumerはfeed-forward behavior-cloning modelを想定してよいが、特定hidden size、activation、loss、optimizerにschemaをbindしない。

---

# Implementation requirements

API名とmodule配置はcurrent Arena architectureに合わせてよいが、次を満たす。

- immutable semantic feature value
- `FEATURE_SEMANTICS_ID`相当のstable identity
- fixed `FEATURE_DIM`
- pure-Python `tensor_values(...)->tuple[float, ...]`相当
- lazy torch importによるoptional `to_tensor(...)`
- tensor dtypeを明示
- feature group metadataまたは文書上のexact index table
- output lengthのruntime assertion
- non-finite valueの拒否
- unsupported schema versionの拒否
- exact input type boundary
- silent clipping / truncation / fallbackなし
- current Arenaのoptional ML dependency boundaryを維持
- existing HandBelief feature / artifact identityを変更しない
- existing action vocabulary / legal-mask contractを変更しない
- current `PolicyInput`型を変更しない
- `lisjong`、`lisjong-engine`へ変更を加えない

normalization scaleは観測されたdataset maximumを根拠に決めない。semantic / rule boundまたは明示的なstable constantを使う。

scaleを超える値を受け取った場合の挙動を明記し、黙ってclipしない。

---

# Schema stability

同一versionでは次を固定する。

- total dimension
- feature group order
- every group offset / length
- logical shape / flatten order
- categorical axis order
- seat mapping
- tile mapping
- presence / padding semantics
- normalization constants
- individual index semantics

全indexのsemantic descriptorからfingerprintを生成し、`version -> expected fingerprint`をtest側のliteralとして固定する。

testは可能な限りimplementation内部のindex builderをsemantic serializerとして再利用しない。

主要block boundaryはhuman-readableなliteral anchorとしても固定する。

互換性を壊す変更では既存fingerprintを書き換えず、新versionを追加する。

ただし、このversioningはArena experiment artifactの再現性のためのものであり、v1を`lisjong`の永久production schemaとして宣言するものではない。

---

# Required tests

## Contract / layout

- exact feature semantics ID
- exact feature dimension
- group rangesにoverlap / holeがない
- every indexに一意なsemantic descriptorがある
- main group boundary anchors
- schema fingerprint
- unsupported version

## Full PolicyInput coverage

current `PolicyInput`の全fieldについて、少なくとも1つのpositive testを持つ。

- self seat
- round wind
- hand number
- dealer
- honba
- riichi sticks
- dora indicators
- live wall tiles remaining
- four players' scores
- discards
- discard tile / red / tsumogiri / order / called_by
- meld kind / tiles / from_seat / called_tile
- riichi state
- own concealed tiles
- drawn tile present / absent

## Determinism / invariance

- same input -> same semantic feature / float tuple
- output length fixed
- all values finite
- all values within documented domains
- self-seat rotation metamorphic test
- unordered tile multiset permutation does not change output
- ordered discard / dora sequence changes the documented positions
- external mutable stateを参照しない

## Boundary cases

- normal five / red five
- drawn tile None / present
- all five meld kinds
- all riichi states
- called / uncalled discard
- minimum / maximum supported sequence lengths
- optional valueとzero valueの区別
- negative scoreをcurrent contractが許容する場合のrepresentation
- West / North round wind
- non-zero honba / riichi sticks
- boundary normalization values

## Fail closed

- wrong input type
- `PolicyInput` subclass
- field subclass等、意味不明な拡張を受理しない方針を採る場合のnegative test
- variable fieldがdocumented maximumを超える
- invalid cross-field sequence that encoder must rely on
- non-finite / out-of-domain numeric value
- feature dimension drift
- unsupported version

## Non-interference / leakage

- encoder public APIが`PolicyInput`以外を要求しない
- `DecisionContext` / legal actionsを要求しない
- hidden truth型を受理しない
- raw engine Observationを受理しない
- phase6 HandBelief feature identity / dimension / artifact testsが変わらない
- `lisjong #149` action vocabulary / legal mask testsが変わらない
- existing Arena evaluation pathsへ影響しない

---

# Documentation

最低限、以下を更新する。

- new experiment-local feature schema document
- `docs/architecture.md`
- public module / usage locationに対応する案内

schema documentには以下を含める。

- scope and ownership
- exact feature semantics ID
- exact dimension
- complete feature group / index table
- logical shapes / flatten order
- axis orders
- seat-relative mapping
- tile / red-five representation
- sequence / padding / presence semantics
- normalization constants
- value domains
- fail-closed conditions
- schema fingerprint / version bump rule
- `PolicyInput`、action vocabulary、legal maskとの関係
- phase6 HandBelief featureとの非互換・非継承関係
- tensor化はadapter責務であること
- experiment-localでありproduction昇格は別Decisionであること

---

# Non-goals

本Issueでは次を実装しない。

- model architecture
- policy network / logits head
- training loop
- loss / optimizer
- behavior-cloning dataset
- teacher action generation
- checkpoint / model artifact
- inference Policy
- legal-mask適用
- action sampling / argmax
- action decode / execution integration
- throughput benchmark
- strength evaluation
- self-play / RL
- HandBelief feature / head / consumer
- hidden-history encoder
- recurrent / Transformer input
- `PolicyInput` schema extension
- generic multi-model feature framework
- `lisjong`へのproduction schema昇格

---

# Stop / re-scope conditions

preflightで次のいずれかが判明した場合は、推測で実装せず、blocker・選択肢・推奨分割を報告する。

- current `PolicyInput`の意味を保ったままfixed-size化できない
- variable-length fieldの安全な上限を確定できず、truncationが必要
- model architectureを先に決めないとrepresentationを選べない
- current `PolicyInput`に最初のBCで不可欠な観測が不足している
- ruleset identityなしではfeature semanticsが曖昧になる
- seat rotationで保持すべき意味が一意に定まらない
- featureがhidden / omniscient stateを必要とする
- existing phase6 schemaをbreaking changeしなければ実装できない
- `lisjong`側contract変更が必要
- feature-only PRではreview可能な一貫した成果物にならない

この場合も、scopeをmodel / trainingまで自動拡張しない。

---

# Acceptance criteria

- [ ] current `PolicyInput`全fieldをpreflight inventoryした
- [ ] PolicyInput-onlyのexperiment-local semantic featureを実装した
- [ ] fixed-size float layoutとoptional tensor adapterを実装した
- [ ] exact version / dimension / index layout / normalizationを文書化した
- [ ] seat-relative semanticsをmetamorphic testで固定した
- [ ] tile / red-five semanticsを固定した
- [ ] ordered sequenceとunordered multisetを正しく区別した
- [ ] padding / presenceがvalid zeroと衝突しない
- [ ] variable-length inputをsilent truncateしない
- [ ] schema fingerprintとboundary anchorsでnumeric assignmentを固定した
- [ ] malformed / unsupported inputをfail closedにした
- [ ] hidden information / legal actions / external objectsを要求しない
- [ ] existing phase6 HandBelief schemaとaction vocabularyへ非干渉
- [ ] model / training / artifactをscopeへ混ぜていない
- [ ] focused tests、Ruff、GitHub Actionsが成功した
- [ ] documentationとimplementationが一致した

---

# Expected follow-up

本Issue完了後、別のbounded Arena Issueで最小behavior-cloning vertical sliceを扱う。

想定する後続scope:

```text
first-party teacher decisions
    -> versioned train/validation dataset
    -> this feature schema
    -> minimal feed-forward policy model
    -> 802 action logits
    -> legal mask
    -> canonical action resolve
```

後続Issueでは少なくとも、

- teacher / population identity
- split / leakage boundary
- artifact metadataにfeature schema versionとaction vocabulary versionをbinding
- illegal selected action 0
- action-family別metric
- end-to-end inference throughput

を扱う。

本Issueのschemaをproduction contractへ昇格するかは、vertical sliceの結果を得た後に別途判断する。
